From 8605ac763e9787a4c3fd57a8d363242c2145554a Mon Sep 17 00:00:00 2001
From: Pascal Paillet <p.paillet@foss.st.com>
Date: Thu, 16 Jan 2025 14:32:51 +0100
Subject: [PATCH 1530/1695] plat-stm32mp1: stm32mp1_pwr: retrieve power domain
 state from EXTI

Instead of using proprietary genPD SMC method,
Get power domain state from the status of the EXTI IMR registers.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 5db328253e73f1f71e6d3c2bcae1e14a6866a2bc plat-stm32mp1: stm32mp1_pwr: retrieve power domain state from EXTI]
Change-Id: Id4ab669a86728845b486aef6a5884031c83a2f0f
Signed-off-by: Pascal Paillet <p.paillet@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/429401
Reviewed-by: Pascal PAILLET <p.paillet@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Tested-by: Pascal PAILLET <p.paillet@st.com>
Reviewed-by: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
---
 .../nsec-service/stm32mp1_svc_setup.c         |   2 +
 core/arch/arm/plat-stm32mp1/pm/context.h      |   2 +
 core/arch/arm/plat-stm32mp1/pm/power_config.c | 179 +++++++++++++++++-
 3 files changed, 176 insertions(+), 7 deletions(-)

diff --git a/core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_svc_setup.c b/core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_svc_setup.c
index f7327be36..481eb6596 100644
--- a/core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_svc_setup.c
+++ b/core/arch/arm/plat-stm32mp1/nsec-service/stm32mp1_svc_setup.c
@@ -33,9 +33,11 @@ static enum sm_handler_ret sip_service(struct sm_ctx *ctx __unused,
 	case STM32_SIP_SVC_FUNC_PWR:
 		args->a0 = pwr_scv_handler(args->a1, args->a2, args->a3);
 		break;
+#ifdef CFG_STM32_LOWPOWER_SIP
 	case STM32_SIP_FUNC_PD_DOMAIN:
 		args->a0 = pm_domain_scv_handler(args->a1, args->a2);
 		break;
+#endif
 	/* Deprecated function IDs */
 	case STM32_SIP_SVC_FUNC_BSEC:
 	case STM32_SIP_SVC_FUNC_SCMI_AGENT:
diff --git a/core/arch/arm/plat-stm32mp1/pm/context.h b/core/arch/arm/plat-stm32mp1/pm/context.h
index 5e5a3cbd2..43a538976 100644
--- a/core/arch/arm/plat-stm32mp1/pm/context.h
+++ b/core/arch/arm/plat-stm32mp1/pm/context.h
@@ -90,7 +90,9 @@ void stm32mp_pm_wipe_context(void);
 TEE_Result stm32mp_pm_call_bl2_lp_entry(unsigned int soc_mode);
 #endif
 
+#ifdef CFG_STM32_LOWPOWER_SIP
 int stm32mp1_set_pm_domain_state(enum stm32mp1_pm_domain domain, bool status);
+#endif
 
 uint32_t stm32mp1_get_lp_soc_mode(uint32_t psci_mode);
 int stm32mp1_set_lp_deepest_soc_mode(uint32_t psci_mode, uint32_t soc_mode);
diff --git a/core/arch/arm/plat-stm32mp1/pm/power_config.c b/core/arch/arm/plat-stm32mp1/pm/power_config.c
index 365f9ac1e..cce4d7eae 100644
--- a/core/arch/arm/plat-stm32mp1/pm/power_config.c
+++ b/core/arch/arm/plat-stm32mp1/pm/power_config.c
@@ -5,6 +5,7 @@
 
 #include <assert.h>
 #include <config.h>
+#include <drivers/stm32_exti.h>
 #include <drivers/stm32mp_dt_bindings.h>
 #include <kernel/dt.h>
 #include <kernel/boot.h>
@@ -28,13 +29,6 @@ static uint32_t system_off_mode;
 static bool retram_enabled;
 static uint8_t stm32mp1_supported_soc_modes[STM32_PM_MAX_SOC_MODE];
 
-/* Boot with all domains ON */
-static bool stm32mp1_pm_dom[STM32MP1_PD_MAX_PM_DOMAIN] = {
-	[STM32MP1_PD_VSW] = false,
-	[STM32MP1_PD_CORE_RET] = false,
-	[STM32MP1_PD_CORE] = false
-};
-
 bool stm32mp1_is_retram_during_standby(void)
 {
 	return retram_enabled;
@@ -71,6 +65,14 @@ bool need_to_backup_stop_context(unsigned int soc_mode)
 	}
 }
 
+#ifdef CFG_STM32_LOWPOWER_SIP
+/* Boot with all domains ON, false means in use */
+static bool stm32mp1_pm_dom[STM32MP1_PD_MAX_PM_DOMAIN] = {
+	[STM32MP1_PD_VSW] = false,
+	[STM32MP1_PD_CORE_RET] = false,
+	[STM32MP1_PD_CORE] = false
+};
+
 static bool get_pm_domain_state(uint8_t mode)
 {
 	bool res = true;
@@ -94,6 +96,154 @@ int stm32mp1_set_pm_domain_state(enum stm32mp1_pm_domain domain, bool status)
 	return 0;
 }
 
+static void dump_pm_domain_state(uint8_t domain __unused)
+{
+}
+#else /* CFG_STM32_LOWPOWER_SIP */
+
+static struct stm32_exti_pdata *exti;
+#define EXTI_BANK_NR		3U
+
+#ifdef CFG_STM32MP13
+/*
+ * Implementation of  PWR Table 35. Functionalities depending on system
+ * operating mode
+ *
+ * Valid wake-up source from LP-Stop modes are:
+ * USBH  42, 43
+ * OTG   44
+ * ETH   68..71
+ */
+#define IMR1_PD_CORE_MASK 0
+#define IMR2_PD_CORE_MASK (GENMASK_32(12, 10))
+#define IMR3_PD_CORE_MASK (GENMASK_32(7, 4))
+
+/*
+ * Valid wake-up source from LPLV-Stop modes are:
+ * GPIO     0..15
+ * PVD/AVD  16
+ * I2C      21..25
+ * USART    26..33
+ * SPI      36..40
+ * LPTIM    47,48,50,52,53
+ */
+#define IMR1_PD_CORE_RET_MASK (GENMASK_32(15, 0) | BIT(16) | \
+			       GENMASK_32(25, 21) | GENMASK_32(31, 26))
+#define IMR2_PD_CORE_RET_MASK (GENMASK_32(8, 4) | GENMASK_32(16, 15) | \
+			       BIT(18) | GENMASK_32(21, 20))
+#define IMR3_PD_CORE_RET_MASK 0
+
+#endif /* CFG_STM32MP13 */
+
+#ifdef CFG_STM32MP15
+/*
+ * Implementation of  Table 34. Functionalities depending on system
+ * operating mode
+ *
+ * Valid wake-up source from LP-Stop mode are:
+ * I2C   21..25
+ * USART 26..33
+ * SPI   36..41
+ * MDIOS 42
+ * USBH  43
+ * OTG   44
+ * LPTIM 47,48
+ * I2C6  54
+ * IPCC  61,62
+ * HSEM  63,64
+ * SEV   65,66
+ * HDMI  69
+ * ETH   70,71
+ * DTS   72
+ * CPU2  73
+ * CDBG  75
+ */
+#define IMR1_PD_CORE_MASK (GENMASK_32(31, 21))
+#define IMR2_PD_CORE_MASK (GENMASK_32(1, 0) | GENMASK_32(16, 4) | BIT(18) | \
+			   GENMASK_32(22, 20) | GENMASK_32(31, 29))
+#define IMR3_PD_CORE_MASK (GENMASK_32(1, 0) | GENMASK_32(9, 4) | BIT(11))
+
+/*
+ * Valid wake-up source from LPLV-Stop modes are:
+ * GPIOs   0..15
+ * PVD/AVD 16
+ */
+#define IMR1_PD_CORE_RET_MASK (GENMASK_32(15, 0) | BIT(16))
+#define IMR2_PD_CORE_RET_MASK 0
+#define IMR3_PD_CORE_RET_MASK 0
+
+#endif /* CFG_STM32MP15 */
+
+static uint32_t imr_pd_core_mask[EXTI_BANK_NR] = {
+	IMR1_PD_CORE_MASK,
+	IMR2_PD_CORE_MASK,
+	IMR3_PD_CORE_MASK,
+};
+
+static uint32_t imr_pd_core_ret_mask[EXTI_BANK_NR] = {
+	IMR1_PD_CORE_RET_MASK,
+	IMR2_PD_CORE_RET_MASK,
+	IMR3_PD_CORE_RET_MASK,
+};
+
+static bool get_domain_state_from_exti(uint8_t domain)
+{
+	unsigned int i = 0;
+
+	for (i = 0; i < EXTI_BANK_NR; i++) {
+		uint32_t imr_mask = imr_pd_core_mask[i];
+		uint32_t imr = stm32_exti_read_imr(exti, i);
+
+		if (domain == STM32MP1_PD_CORE_RET)
+			imr_mask |= imr_pd_core_ret_mask[i];
+
+		if (imr & imr_mask)
+			return true;
+	}
+
+	return false;
+}
+
+/* The function returns FALSE if the domain is in use. */
+static bool get_pm_domain_state(uint8_t domain)
+{
+	if (!exti)
+		return true;
+
+	return get_domain_state_from_exti(domain) == false;
+}
+
+static void dump_pm_domain_state(uint8_t __maybe_unused domain)
+{
+#ifdef POWER_DEBUG
+	unsigned int i = 0;
+
+	for (i = 0; i < EXTI_BANK_NR; i++) {
+		uint32_t imr = stm32_exti_read_imr(exti, i);
+		uint32_t imr_mask = imr_pd_core_mask[i];
+		const char *name = "PD_CORE";
+
+		if (domain == STM32MP1_PD_CORE_RET) {
+			imr_mask = imr_pd_core_ret_mask[i];
+			name = "PD_CORE_RET";
+		}
+
+		if (imr & imr_mask) {
+			unsigned int bit = 0;
+
+			/* Log EXTI numbers using domain */
+			for (bit = 0; bit < 32; bit++) {
+				if ((imr & imr_mask) & BIT(bit))
+					IMSG("Domain %s needed for EXTI %u",
+					     name, i * U(32) + bit);
+			}
+		}
+	}
+#endif
+}
+
+#endif  /* CFG_STM32_LOWPOWER_SIP */
+
 #ifdef CFG_EMBED_DTB
 static void save_supported_mode(void *fdt, int pwr_node)
 {
@@ -140,11 +290,16 @@ uint32_t stm32mp1_get_lp_soc_mode(uint32_t psci_mode)
 
 	mode = deepest_suspend_mode;
 
+	dump_pm_domain_state(STM32MP1_PD_CORE);
+	dump_pm_domain_state(STM32MP1_PD_CORE_RET);
+
+	/* if PD_CORE_RET is in use don't allow deeper than Standby */
 	if (mode == STM32_PM_CSTOP_ALLOW_STANDBY_DDR_SR &&
 	    (!get_pm_domain_state(STM32MP1_PD_CORE_RET) ||
 	     !is_supported_mode(mode)))
 		mode = STM32_PM_CSTOP_ALLOW_LPLV_STOP2;
 
+	/* if PD_CORE is in use don't allow deeper than LPLV-Stop */
 	if (mode == STM32_PM_CSTOP_ALLOW_LPLV_STOP2 &&
 	    (!get_pm_domain_state(STM32MP1_PD_CORE) ||
 	     !is_supported_mode(mode)))
@@ -198,6 +353,7 @@ static TEE_Result stm32mp1_init_lp_states(void)
 	void *fdt = NULL;
 	int pwr_node = -1;
 	const fdt32_t *cuint = NULL;
+	TEE_Result __maybe_unused res = TEE_ERROR_GENERIC;
 
 	fdt = get_embedded_dt();
 	if (fdt)
@@ -220,6 +376,15 @@ static TEE_Result stm32mp1_init_lp_states(void)
 	/* Initialize suspend support to the deepest possible mode */
 	deepest_suspend_mode = STM32_PM_CSTOP_ALLOW_STANDBY_DDR_SR;
 
+#ifndef CFG_STM32_LOWPOWER_SIP
+	res = dt_driver_device_from_node_idx_prop("wakeup-parent",
+						  fdt, pwr_node, 0,
+						  DT_DRIVER_INTERRUPT,
+						  &exti);
+	if (res)
+		return res;
+#endif
+
 #ifdef CFG_STM32MP1_OPTEE_IN_SYSRAM
 	if (!stm32mp_supports_hw_cryp())
 		deepest_suspend_mode = STM32_PM_CSTOP_ALLOW_LPLV_STOP;
-- 
2.47.2

